Content.Server/Body/Systems/BodyReassembleSystem.cs

```diff
@@ -1,229 +1,279 @@
 
-            // Custom verb
-            AlternativeVerb custom = new()
+        // Custom verb
+        AlternativeVerb custom = new()
+        {
+            Text = Loc.GetString("reassemble-action"),
+            Act = () =>
             {
-                Text = Loc.GetString("reassemble-action"),
-                Act = () =>
-                {
-                    StartReassembly(uid, component, multiplier);
-                },
-                IconEntity = uid,
-                Priority = 1
-            };
-            args.Verbs.Add(custom);
-        }
+                StartReassembly(uid, component, multiplier);
+            },
+            IconEntity = uid,
+            Priority = 1
+        };
+        args.Verbs.Add(custom);
+    }
 
-        private bool GetNearbyParts(EntityUid uid, BodyReassembleComponent component, out HashSet<EntityUid>? partList)
+    private bool GetNearbyParts(EntityUid uid, BodyReassembleComponent component, out HashSet<EntityUid>? partList)
+    {
+        partList = new HashSet<EntityUid>();
+        
+        if (component.BodyParts == null)
+            return false;
+
+        var tempList = new HashSet<EntityUid>(component.BodyParts);
+        tempList.Remove(component.Owner);
+        partList.Add(component.Owner); //the query dislikes the owner
+
+        var entq = GetEntityQuery<BodyPartComponent>();
+        EntityUid? foundPart = null;
+        foreach (var entity in _lookup.GetEntitiesInRange(component.Owner, 2f))
         {
-            partList = new HashSet<EntityUid>();
+            if (!entq.HasComponent(entity))
+                continue;
 
-            if (component.BodyParts == null)
-                return false;
+            foreach (var part in tempList)
+            {
+                var partmeta = MetaData(part);
+                var entmeta = MetaData(entity);
 
-            // Ensures all of the old body part pieces are there
-            var xformQuery = GetEntityQuery<TransformComponent>();
-            var bodyXform = xformQuery.GetComponent(uid);
+                if (partmeta.EntityPrototype == null || entmeta.EntityPrototype == null)
+                    continue;
 
-            foreach (var part in component.BodyParts)
-            {
-                if (!xformQuery.TryGetComponent(part, out var xform) ||
-                    !bodyXform.Coordinates.InRange(EntityManager, xform.Coordinates,2f))
+                if (partmeta.EntityPrototype.ID == entmeta.EntityPrototype.ID)
                 {
-                    _popupSystem.PopupEntity(Loc.GetString("reassemble-fail"), uid, Filter.Entities(uid));
-                    return false;
+                    partList.Add(entity);
+                    foundPart = part;
                 }
-                partList.Add(part);
             }
-            return true;
+
+            if (foundPart != null)
+                tempList.Remove(foundPart.Value);
+
+            foundPart = null;
```

Comment:
This is dirty as, please don't add the entitylookup back in and especially don't compare prototype IDs.